---
title: "GTEx single cell data"
subtitle: '[Eraslan, et al. biorxiv (2021)](https://www.biorxiv.org/content/10.1101/2021.07.19.452954v1)'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
vignette: >
  %\VignetteIndexEntry{GTEx}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---


<!---

cd /sc/arion/projects/CommonMind/hoffman/crumblr_analysis/GTEx
ml python git
git pull
R --vanilla


system("git pull"); rmarkdown::render("1M-scBloodNL_crumblr.Rmd");


https://hoffmg01.hpc.mssm.edu/crumblr_analysis/1M-scBloodNL/1M-scBloodNL_crumblr.html


--->

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  cache = TRUE,
  cache.lazy = FALSE)
```

```{r get.data, eval=FALSE}
# https://gtexportal.org/home/datasets

cd /hpc/users/hoffmg01/work/eval_methods/dreamlet/data

wget --no-check-certificate https://storage.googleapis.com/gtex_analysis_v9/snrna_seq_data/GTEx_8_tissues_snRNAseq_atlas_071421.public_obs.h5ad
wget --no-check-certificate https://storage.googleapis.com/gtex_analysis_v9/snrna_seq_data/GTEx_8_tissues_snRNAseq_immune_atlas_071421.public_obs.h5ad
```

```{r load.packages, cache=FALSE}
library(zellkonverter)
library(SingleCellExperiment)
library(dreamlet)
library(qvalue)
library(tidyverse)
library(zenith)
library(gplots)
```

```{r load.data}
# specify HDF5 file
file = '/hpc/users/hoffmg01/work/eval_methods/dreamlet/data/GTEx_8_tissues_snRNAseq_atlas_071421.public_obs.h5ad'

sce = readH5AD( file, use_hdf5=TRUE)

# remove scrublet: True
sce = sce[,colData(sce)$scrublet=='False']
```


# Joint UMAP
```{r umap, dev="png", fig.height=12, fig.width=8}
# extract UMAP coordinates and annotations
df = cbind(reducedDim(sce, "X_umap"), colData(sce))
df = data.frame(df)

ggplot(df, aes(V1, V2, color=Granular.cell.type)) + geom_point(size=.05) + theme_classic() + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="bottom") + guides(colour = guide_legend(override.aes = list(size = 1.5))) + xlab("UMAP1") + ylab("UMAP2") 
```



```{r aggregateToPseudoBulk}
# Create pseudo-bulk SingleCellExperiment
pbObj <- aggregateToPseudoBulk(sce,
    assay = "counts", 
    cluster_id = "Granular.cell.type",
    sample_id  = "channel")
```

```{r canCorPairs}
# RIN.score.from.PAXgene.tissue.Aliquot
# RIN.score.from.Frozen.tissue.Aliquot
form = ~ tissue + prep + Participant.ID + Age_bin + Sex +  RIN.score.from.PAXgene.tissue.Aliquot + Sample.Ischemic.Time..mins. # Autolysis.Score

C = canCorPairs( form, colData(pbObj) )

plotCorrMatrix(C)
```

```{r processAssays}
form = ~ (1|tissue) + (1|prep) + (1|Participant.ID) + (1|Age_bin) + (1|Sex) +  RIN.score.from.PAXgene.tissue.Aliquot + Sample.Ischemic.Time..mins. # Autolysis.Score

# include cell types with > 200K reads
totalReads = sapply(assayNames(pbObj), function(x) sum(assay(pbObj, x)))

vObj = processAssays(pbObj, form,
	min.samples = 10,
    BPPARAM = SnowParam(12, progressbar=TRUE))
```


```{r plotVoom, fig.height=18, fig.width=9}
plotVoom(vObj, ncol=4)
```

```{r vp}
vp = fitVarPart( vObj, form,
    BPPARAM = SnowParam(12, progressbar=TRUE))
```

```{r plot.vp, fig.height=25, fig.width=12}
plotVarPart(vp, label.angle=80)
```




```{r dreamlet, eval=FALSE}
form = ~ prep + (1|tissue) + (1|Participant.ID) + (1|Age_bin) + (1|Sex) +  RIN.score.from.PAXgene.tissue.Aliquot + Autolysis.Score + Sample.Ischemic.Time..mins.

# define contrast formulas
lvls = levels(colData(res.proc)$prep)
lvls = paste0('prep', lvls)

grid = expand.grid(lvls,lvls)
idx = apply(grid, 1, function(x) length(unique(x)))
grid = grid[idx==2,]

ids = apply(grid, 1, function(x){
	paste0(x[1], '_', x[2])
	})

formulas = apply(grid, 1, function(x){
	paste0(x[1], ' - ', x[2])
	})

names(formulas) = ids

fit = dreamlet(vObj, form,
	# contrasts = formulas,
    BPPARAM = SnowParam(12, progressbar=TRUE))
```



Myocyte UMAP
```{r umap, dev="png", fig.height=12, fig.width=8}
# extract UMAP coordinates and annotations
df = cbind(reducedDim(sce, "X_umap"), colData(sce))
df = data.frame(df)
keep = grep("Myocyte", df$Granular.cell.type)

ggplot(df[keep,], aes(V1, V2, color=Granular.cell.type)) + geom_point(size=.05) + theme_classic() + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="bottom") + guides(colour = guide_legend(override.aes = list(size = 1.5))) + xlab("UMAP1") + ylab("UMAP2") 
```

```{r dreamletCompareClusters}

assayNames(pbObj)[grep("Myocyte", assayNames(pbObj))]


ct1 = "Myocyte (fast-twitch, cytoplasmic)"
ct2 = "rest"
fitCompare = dreamletCompareClusters(pbObj, 
  assays = c(ct1, ct2),
  min.cells = 2,
  min.count = 5,
  min.samples = 3,
  method = "random",
  formula = ~ (1|prep),
  BPPARAM = SnowParam(12, progressbar=TRUE) )
```

```{r topTable}
tab = topTable(fitCompare, coef="compare", number=Inf)
table(tab$adj.P.Val < 0.05)

plotVolcano(fitCompare, coef="compare")

tabTr = getTreat(fitCompare, lfc = log2(2), coef="compare", number=Inf)
table(tabTr$adj.P.Val < 0.05)

df = merge(tab, tabTr, by="row.names")
with(df, cor(t.x, t.y, method="sp" ))
```


```{r zenith, fig.heigth=12, fig.width=7}
go.gs = get_GeneOntology(to="SYMBOL")
   
# Run zenith gene set analysis on result of dreamlet
res_zenith = zenith_gsa(fitCompare, go.gs, 'compare' )

plotZenithResults(res_zenith, 30, 20)
```




```{r exit1, cache=FALSE, echo=FALSE}
knitr::knit_exit()
```

# consider analysis within tissue


# Cell type composition
```{r crumblr}
library(crumblr)
library(cowplot)

idx = !(pbObj$tissue %in% c("lung", "skeletalmuscle"))
# idx[] = TRUE

counts = cellCounts(pbObj[,idx])

fractions = apply(counts, 1, function(x) (x+.5) / sum(x+0.5))

cobj = crumblr(counts)
info_sub = colData(pbObj[,idx])


f = function( method ){
  if( method == "vst" ){
    # Apply variance stabilizing transform
    M = vst(cobj)
    title = "VST"
  }else if( method == "CLR"){
    M = cobj$E
    title = "CLR"
  }else{
    M = fractions
    title = "Fractions"
  }

  if( TRUE ){ 
    # Perform PCA
    pca = prcomp(t(M))

    # # merge with metadata 
    df_pca = merge(pca$x, info_sub, by="row.names")

    frac = pca$sdev^2 / sum(pca$sdev^2)
    frac = format(100*frac[1:3], digits=2)
  }else{
    df_cmd = cmdscale(dist(t(M)), k=2)
    colnames(df_cmd) = c("PC1", "PC2")
    frac = c(0,0,0)
    df_pca = merge(df_cmd, info_sub, by="row.names")
  }

  ggplot(df_pca, aes(PC1, PC2, color=tissue, shape=prep)) + 
          geom_point(size=3) + 
          theme_classic() + 
          theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + 
          scale_color_discrete(name="Subject") +
          xlab(paste0("PC1 (", frac[1], '%)')) + ylab(paste0("PC2 (", frac[2], '%)')) + ggtitle(title)
}

fig = plot_grid(f("frac"), 
  f("CLR") + theme(legend.position="none"), 
  f("vst") + theme(legend.position="none"), 
  align="hv", axis="tblr", nrow=1)

file = "/sc/arion/projects/CommonMind/hoffman/NPS-AD/work/nps_ad/test.png"
ggsave(file, fig, height=18, width=15)  

png(file, width=1200, height=800)
par(mar=c(7,4,4,2)+0.1) 
heatmap.2((cobj$E), trace = "none", density.info = "none", col = bluered(100) , keysize=1, cexRow=1, cexCol=1, margins=c(12,8))
dev.off()


library(nnet)

methods = c("frac", "CLR", "vst")

df = lapply(methods, function(method){

   if( method == "vst" ){
    # Apply variance stabilizing transform
    M = vst(cobj)
  }else if( method == "CLR"){
    M = cobj$E
  }else{
    M = fractions
  }
  pca = prcomp(t(M))
  df_pca = merge(pca$x, info_sub, by="row.names")

  form = tissue ~ 1 

  df = lapply( 1:5, function(k){
    form <<- update.formula(form, paste0(". ~ .+ PC", k))

    fit = multinom(form, droplevels(df_pca), trace=FALSE)
    data.frame(method = method, k = k, deviance = fit$deviance)
  })
  do.call(rbind, df)
})
df = do.call(rbind, df)


fig = ggplot(df, aes(k, deviance, color=method)) +
  geom_line()  +
  theme_classic()
  theme(aspect.ratio=1)

ggsave(file, fig)  




```

```{r fracs, fig.height=12}
plotPercentBars(data.frame(t(fractions)))
```

```{r vp.crumblr}
cobj = crumblr(cellCounts(pbObj))

form = ~ (1|tissue) + (1|prep) + (1|Participant.ID)
vp.clr = fitExtractVarPartModel( cobj$E, form, colData(pbObj))

vp = fitExtractVarPartModel( cobj, form, colData(pbObj))

# png(file)
fig1 = plotVarPart(sortCols(vp.clr)) + ggtitle("CLR")
fig2 = plotVarPart(sortCols(vp))  + ggtitle("CLR + weights")
plot_grid(fig1, fig2, nrow=1)
# dev.off()
```

```{r cca, eval=FALSE}
P1 = "CST"
P2 = "TST"

i1 = which( colData(pbObj)$prep == P1)
i2 = which( colData(pbObj)$prep == P2)

info = merge(colData(pbObj)[i1,], colData(pbObj)[i2,], by=c("tissue", "Participant.ID"))

j1 = which( colData(pbObj)$tissue.individual.prep %in% info$tissue.individual.prep.x)
j2 = which( colData(pbObj)$tissue.individual.prep %in% info$tissue.individual.prep.y)

cobj = crumblr(cellCounts(pbObj))

X1 = t(vst(cobj[,j1]))
X2 = t(vst(cobj[,j2]))

decorrelate::fastcca(X1, X2, lambda.x=.2, lambda.y=.2)

X1 = t(cobj[,j1]$E)
X2 = t(cobj[,j2]$E)

decorrelate::fastcca(X1, X2, lambda.x=.2, lambda.y=.2)
```


















